crypto/ec/curve448/constant_time.h

```diff
@@ -0,0 +1,337 @@
+ * a change from the Ed448 versions.
+ *
+ * These routines would be much simpler to define at the byte level,
+ * but if not vectorized they would be a significant fraction of the
+ * runtime.  Eg on NEON-less ARM, constant_time_lookup is like 15% of
+ * signing time, vs 6% on Haswell with its fancy AVX2 vectors.
+ *
+ * If the compiler could do a good job of autovectorizing the code,
+ * we could just leave it with the byte definition.  But that's unlikely
+ * on most deployed compilers, especially if you consider that pcmpeq[size]
+ * is much faster than moving a scalar to the vector unit (which is what
+ * a naive autovectorizer will do with constant_time_lookup on Intel).
+ *
+ * Instead, we're putting our trust in the loop unroller and unswitcher.
+ */
+
+# if defined(__GNUC__) || defined(__clang__)
+/*
+ * Unaligned big (vector?) register.
+ */
+typedef struct {
+    big_register_t unaligned;
+} __attribute((packed)) unaligned_br_t;
+
+/*
+ * Unaligned word register, for architectures where that matters.
+ */
+typedef struct {
+    word_t unaligned;
+} __attribute((packed)) unaligned_word_t;
+
+#  define HAS_UNALIGNED_STRUCTS
+#  define RESTRICT __restrict__
+#else
+#  define RESTRICT
+# endif
+
+/*
+ * Constant-time conditional swap.
+ *
+ * If doswap, then swap elem_bytes between *a and *b.
+ *
+ * *a and *b must not alias.  Also, they must be at least as aligned
+ * as their sizes, if the CPU cares about that sort of thing.
+ */
+static ossl_inline void constant_time_cond_swap(void *RESTRICT a_,
+                                                void *RESTRICT b_,
+                                                word_t elem_bytes,
+                                                mask_t doswap)
+{
+    word_t k;
+    unsigned char *a = (unsigned char *)a_;
+    unsigned char *b = (unsigned char *)b_;
+    big_register_t br_mask = br_set_to_mask(doswap);
+# ifndef HAS_UNALIGNED_STRUCTS
+    unsigned char doswapc = (unsigned char)(doswap & 0xFF);
+# endif
+
+    for (k = 0; k <= elem_bytes - sizeof(big_register_t);
+         k += sizeof(big_register_t)) {
+        if (elem_bytes % sizeof(big_register_t)) {
+            /* unaligned */
+# ifdef HAS_UNALIGNED_STRUCTS
+            big_register_t xor = ((unaligned_br_t *) (&a[k]))->unaligned
+                                 ^ ((unaligned_br_t *) (&b[k]))->unaligned;
+
+            xor &= br_mask;
+            ((unaligned_br_t *)(&a[k]))->unaligned ^= xor;
+            ((unaligned_br_t *)(&b[k]))->unaligned ^= xor;
+# else
+            size_t i;
+
+            for (i = 0; i < sizeof(big_register_t); i++) {
+                unsigned char xor = a[k + i] ^ b[k + i];
+
+                xor &= doswapc;
+                a[k + i] ^= xor;
+                b[k + i] ^= xor;
+            }
+# endif
+        } else {
+            /* aligned */
+            big_register_t xor = *((big_register_t *) (&a[k]))
+                                 ^ *((big_register_t *) (&b[k]));
```

Comment:
Hmm....

big_register_t is likely not the effective type of the object pointed to with a, and b.
Thus this seems to be in violation of the strict-aliasing rules.

Reply:
Hmmm. Nasty - could be tricky to fix. I decided to bite the bullet and rip out all the curve448 specific constant time stuff and use our standard calls. It may result in some performance degradation, but I guess we worry about that later.